status/status.go

```diff
@@ -128,3 +130,39 @@ func FromError(err error) (s *Status, ok bool) {
 	}
 	return nil, false
 }
+
+// WithDetails returns a new status with the provided details messages appended to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) WithDetails(details ...proto.Message) (*Status, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.Proto() != nil.
+	p := s.Proto()
+	for _, detail := range details {
+		any, err := ptypes.MarshalAny(detail)
+		if err != nil {
+			return nil, err
+		}
+		p.Details = append(p.Details, any)
+	}
+	return &Status{s: p}, nil
+}
+
+// Details returns a slice of details messages attached to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) Details() ([]proto.Message, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.s != nil.
+	details := make([]proto.Message, 0, len(s.s.Details))
+	for _, any := range s.s.Details {
+		detail := &ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(any, detail); err != nil {
+			return nil, err
+		}
```

Comment:
FWIW, I wrote most of this yesterday before @ghasemloo's reply, but did not finish before leaving the office.  tl;dr: let's go with 3!

> I kind of like 4b because preserves type-safety,

Not to be pedantic, but all of the solutions proposed are type-safe; this would declare more-specific *static* types than 3, but the actual utility of that extra type specificity is questionable (proto.Message will need to be type asserted to a concrete message type to be used meaningfully anyway).

> handles the common case (people won't care about the error usually), but makes it easy to get the errors if needed (without duplicating the body of Details).

One downside with 4a vs. 1 is that you have to iterate over the results by index instead of entry if you care about the errors.  I'm also somewhat opposed to needing to line up parallel data structures to make sense of things where it can be avoided.

4b is nice in that the errors are hidden until requested, but has some moderate wasted effort to re-parse everything in order to find _only_ the errors.  I would rather have two functions, one that returns only the messages that could be successfully unmarshalled and one that returns the messages and the errors, and different users would call different ones.

Given that there are other errors that come from unmarshalling besides the type being unknown, I'd like to go with only 3 for now, still.  We could later add a `func (s *Status) ValidDetails() []proto.Message` (open to better name suggestions) if there turns out to be a reasonable case for having an easier-to-use version.

A bit of bikeshedding: it's too bad Go doesn't have a native/standardized [`Either`](https://hackage.haskell.org/package/base-4.9.1.0/docs/Data-Either.html) type that could be used instead of a custom `struct` or `interface{}` -- which have poorer semantics -- or else that would be the obvious choice.